# Assignment 1
You should submit the **UniversityNumber.ipynb** file and your final prediction file **UniversityNumber.test.txt** to moodle. Make sure your code does not use your local files and that the results are reproducible. Before submitting, please **1. clean all outputs and 2. run all cells in your notebook and keep all running logs** so that we can check.

# 1 $n$-gram Language Model

In [25]:
!mkdir -p data/lm
!wget -O data/lm/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/train.txt
!wget -O data/lm/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/dev.txt
!wget -O data/lm/test.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/lm/test.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 1.1 Building vocabulary

### Code

In [16]:
with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
    train = dict()
    for line in f:
        line = '<START> ' + line + ' <END>'     # to add <START> and <END> to the start and end of each sentence 
        tmp = line.split()
        for word in tmp:
            if word in train.keys():            # if in dict, increment counter
                train[word] += 1
            else:
                train[word] = 1                 # if not in dict, add new entry

    countUNK = 0
    for key in list(train.keys()):              # find all words with freq < 3, sum their counts
        if train[key] < 3:
            countUNK += train[key]
            del train[key]
    train['<UNK>'] = countUNK
    print("Size of vocabulary = " + str(len(train)))

Size of vocabulary = 26601


### Discussion

Size of vocabulary = 26601

Number of parameters of n-gram models = (size of vocabulary) ^ n = (26601) ^ n
eg unigram: 26601 ^ 1
eg bigram: 26601 ^ 2
eg trigram: 26601 ^ 3

## 1.2 $n$-gram Language Modeling

### Code

In [22]:
import math

unigramDict = dict()
bigramDict = dict()
bigramTrain = dict()

# build unigram
def unigram():
    for key in train.keys():
        unigramDict[key] = float(train[key]) / sum(train.values())

def unigramProb(word):
    if word in unigramDict.keys():
        return unigramDict[word]
    else:
        return unigramDict['<UNK>']

# build bigram
def bigramVocab():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 1):
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                    
                if (tmp[i], tmp[i+1]) in bigramTrain.keys():
                    bigramTrain[(tmp[i], tmp[i+1])] += 1
                else:
                    bigramTrain[(tmp[i], tmp[i+1])] = 1

def bigram():
    for key in bigramTrain.keys():
        bigramDict[key] = float(bigramTrain[key]) / train[key[0]]    # P(w2|w1) = C(w1, w2) / C(w1)
           
unigram()
bigramVocab()
bigram()


In [23]:
def unigramPerplexity(set):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            # print(tmp)
            logProb = 0
            for word in tmp:
                N += 1
                # logProb *= unigramProb(word)
                logProb += math.log2(unigramProb(word))
                # print(logProb, unigramProb(word), word, line)
            sum += (logProb)
    return 2 ** (-sum / N)

def bigramPerplexity(set):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 0
            for i in range(len(tmp) - 1):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'
                    
                if (tmp[i], tmp[i+1]) not in bigramDict.keys():
                    prob = 0 / train[tmp[i]]
                else:
                    prob = bigramDict[(tmp[i], tmp[i+1])]
                # logProb += math.log2(bigramDict[(tmp[i], tmp[i+1])])
                logProb += math.log2(prob)
            sum += (logProb)
    return 2 ** (-sum / N)

print("Unigram train perplexity = " + str(unigramPerplexity("train")))
print("Unigram dev perplexity = " + str(unigramPerplexity("dev")))
print("Bigram train perplexity = " + str(bigramPerplexity("train")))
print("Bigram dev perplexity = " + str(bigramPerplexity("dev")))


Unigram train perplexity = 888.8280671219554
Unigram dev perplexity = 815.9305409758995
Bigram train perplexity = 76.93455019701061


ValueError: math domain error

### Discussion

When calculating the perplexity of unigram on the dev set, even if the words aren't in the training set, they are counted as \<UNK\>, a valid token in the vocabulary.
However, when it comes to the bigram model, some new combinations of words are in the dev set does not exist in the training set. These combinations have a count and probability of zero, making the log likelihood of the model negative infinity which caused error when calculating the perplexity.

## 1.3 Smoothing

### 1.3.1 Add-one (Laplace) smoothing

### Code

In [24]:
bigramSmoothDict = dict()

def bigramSmoothPerplexity(set, k):
    for key in bigramTrain.keys():
        bigramSmoothDict[key] = (float(bigramTrain[key]) + k) / (train[key[0]] + k*26601)    # P(w2|w1) = (C(w1, w2) + 1)/ (C(w1) + |V|)
    
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 0
            for i in range(len(tmp) - 1):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i+1] not in train.keys():
                    tmp[i+1] = '<UNK>'

                if (tmp[i], tmp[i+1]) not in bigramSmoothDict.keys():
                    prob = (k) / (train[tmp[i]] + k*26601)
                else:
                    prob = bigramSmoothDict[(tmp[i], tmp[i+1])]
                    
                logProb += math.log2(prob)
            sum += (logProb)
    return 2 ** (-sum / N)

# bigramTrain()
print("Bigram train perplexity = " + str(bigramSmoothPerplexity("train", 1)))
print("Bigram dev perplexity = " + str(bigramSmoothPerplexity("dev", 1)))


Bigram train perplexity = 1442.5940705284504
Bigram dev perplexity = 1660.2006059675152


#### Discussion

The perplexity after smoothing is worse than before smoothing in 1.2. This is caused by the nature of add-one smoothing being overly simplistic, shifting too much probability mass from seen to unseen bigrams.

#### 1.3.2 Add-k smoothing

##### Code

In [26]:
print("Bigram train perplexity (k=0.001) = " + str(bigramSmoothPerplexity("train", 0.001)))
print("Bigram train perplexity (k=0.01) = " + str(bigramSmoothPerplexity("train", 0.01)))
print("Bigram train perplexity (k=0.1) = " + str(bigramSmoothPerplexity("train", 0.1)))
print()
print("Bigram dev perplexity (k=0.001) = " + str(bigramSmoothPerplexity("dev", 0.001)))
print("Bigram dev perplexity (k=0.01) = " + str(bigramSmoothPerplexity("dev", 0.01)))
print("Bigram dev perplexity (k=0.1) = " + str(bigramSmoothPerplexity("dev", 0.1)))


Bigram train perplexity (k=0.001) = 94.95746093685625
Bigram train perplexity (k=0.01) = 157.7053131622282
Bigram train perplexity (k=0.1) = 407.4502916262101

Bigram dev perplexity (k=0.001) = 432.4287195649327
Bigram dev perplexity (k=0.01) = 437.56977580286156
Bigram dev perplexity (k=0.1) = 695.6444386896961


##### Discussion

From my result, it seems that the perplexity is lower when k is smaller. When k gets closer to 0, the calculated perplexity is also closer to the perplexity of bigram model without any smoothing, while taking care of unseen bigrams.

### 1.3.3 Linear Interpolation

#### Code

In [32]:
trigramDict = dict()
trigramTrain = dict()
def trigram():
    with open('data/lm/train.txt', 'r', encoding="utf-8") as f:
        for line in f:
            line = '<START> <START> ' + line + ' <END>'
            tmp = line.split()
            for i in range(len(tmp) - 2):
                if (tmp[i], tmp[i-1], tmp[i+2]) in trigramTrain.keys():
                    trigramTrain[(tmp[i], tmp[i+1], tmp[i+2])] += 1
                else:
                    trigramTrain[(tmp[i], tmp[i+1], tmp[i+2])] = 1

    for key in trigramTrain.keys():
        trigramDict[key] = float(trigramTrain[key]) / bigramDict[(key[0], key[1])]    # P(w3|w1, w2) = C(w1, w2, w3) / C(w1, w2)

# l1 * P(w3|w1, w2) + l2 * P(w2|w1) + l3 * P(w1)
def linearPerplexity(set, l1, l2, l3):
    with open(f'data/lm/{set}.txt', 'r', encoding="utf-8") as f:
        N = 0
        sum = 0
        for line in f:
            line = '<START> ' + line + ' <END>'
            tmp = line.split()
            logProb = 0
            for i in range(2, len(tmp)):
                N += 1
                if tmp[i] not in train.keys():
                    tmp[i] = '<UNK>'
                if tmp[i-1] not in train.keys():
                    tmp[i-1] = '<UNK>'
                if tmp[i-2] not in train.keys():
                    tmp[i-2] = '<UNK>'
                    
                # trigram: l1 * P(w3|w1, w2)
                if (tmp[i-2], tmp[i-1], tmp[i]) not in trigramDict.keys():
                    prob = 0
                else:
                    prob = l1 * trigramDict[(tmp[i-2], tmp[i-1], tmp[i])]

                # bigram: l2 * P(w2|w1)
                if (tmp[i-1], tmp[i]) not in bigramDict.keys():
                    prob +=  0
                else:
                    prob += l2 * bigramDict[(tmp[i-1], tmp[i])]
                
                # unigram: l3 * P(w1)
                prob += l3 * unigramProb(tmp[i]) 
                
                logProb += math.log2(prob)

            sum += (logProb)
    return 2 ** (-sum / N)

print("Linear interpolation train perplexity (l1=0.2, l2=0.3, l3=0.5) = " + str(linearPerplexity("train", 0.2, 0.3, 0.5)))
print("Linear interpolation dev perplexity (l1=0.2, l2=0.3, l3=0.5) = " + str(linearPerplexity("dev", 0.2, 0.3, 0.5)))

Linear interpolation train perplexity (l1=0.2, l2=0.3, l3=0.5) = 179.4409002188491
Linear interpolation dev perplexity (l1=0.2, l2=0.3, l3=0.5) = 403.89796582139195


In [33]:
bestPerplexity = float('inf')
bestL1 = 0
bestL2 = 0
bestL3 = 0
for i in range(0, 11):
    for j in range(0, 11):
        for k in range(1, 11):
            if i + j + k == 10:
                l1 = i/10
                l2 = j/10
                l3 = k/10
                
                perplexity = linearPerplexity("dev", l1, l2, l3)
                print("Linear interpolation dev perplexity (l1=" + str(l1) + ", l2=" + str(l2) + ", l3=" + str(l3) + ") = " + str(perplexity))
                if perplexity < bestPerplexity:
                    bestPerplexity = perplexity
                    bestL1 = l1
                    bestL2 = l2
                    bestL3 = l3
                    
print("Best l1, l2, l3 = " + str(bestL1) + ", " + str(bestL2) + ", " + str(bestL3))
print("Best dev perplexity = " + str(bestPerplexity))
      

Linear interpolation dev perplexity (l1=0.0, l2=0.0, l3=1.0) = 902.7562574324129
Linear interpolation dev perplexity (l1=0.0, l2=0.1, l3=0.9) = 487.83307571311514
Linear interpolation dev perplexity (l1=0.0, l2=0.2, l3=0.8) = 395.8292004906807
Linear interpolation dev perplexity (l1=0.0, l2=0.3, l3=0.7) = 346.8932206204395
Linear interpolation dev perplexity (l1=0.0, l2=0.4, l3=0.6) = 316.8297990611647
Linear interpolation dev perplexity (l1=0.0, l2=0.5, l3=0.5) = 297.85260261819815
Linear interpolation dev perplexity (l1=0.0, l2=0.6, l3=0.4) = 286.9845930678361
Linear interpolation dev perplexity (l1=0.0, l2=0.7, l3=0.3) = 283.7119180501348
Linear interpolation dev perplexity (l1=0.0, l2=0.8, l3=0.2) = 290.39149030698076
Linear interpolation dev perplexity (l1=0.0, l2=0.9, l3=0.1) = 318.7553419510995
Linear interpolation dev perplexity (l1=0.1, l2=0.0, l3=0.9) = 1003.0625082582239
Linear interpolation dev perplexity (l1=0.1, l2=0.1, l3=0.8) = 526.2226609557068
Linear interpolation dev

#### Discussion

I tried all combinations of hyperparameters where λ1, λ2, λ3 are upto 1 decimal place.
The best hyperparameter set I got is λ1 = 0.0, λ2 = 0.7, λ3 = 0.3, with a perplexity of 283.7119180501348.
However, I do not rule out the possiblity that other sets of hyperparameters that go beyond 1 decimal place can achieve lower perplexity. If that is the case, there would be infintely more set that would need to be tested for further optimisation.

##### 1.3.4 Optimization

#### Discussion

Bayesian optimization could help find the optimal set of hyperparameters. 

# 2 Word Vectors

In [ ]:
def load_embedding_model():
    """ Load GloVe Vectors
        Return:
            wv_from_bin: All 400000 embeddings, each lengh 50
    """
    import gensim.downloader as api
    wv_from_bin = api.load("glove-wiki-gigaword-200")
    print("Loaded vocab size %i" % len(list(wv_from_bin.index_to_key)))
    return wv_from_bin

wv_from_bin = load_embedding_model()

Loaded vocab size 400000


## 2.1 Find most similar word
Use cosine similarity to find the most similar word to each of these words. Report the most similar word and its cosine similarity.

In [ ]:
import numpy as np
from numpy.linalg import norm

wordList = ["dog", "whale", "before", "however", "fabricate"]
wordDict = {}
vecKeys = wv_from_bin.index_to_key
for word in wordList:
  sim = -1
  most = ""
  for i in range(len(wv_from_bin)):
    x = wv_from_bin[word]
    cosine = np.dot(x,wv_from_bin[i])/(norm(x)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    if cosine >= sim and curword != word:
      sim = cosine
      most = curword
  wordDict[word] = [most, sim]
print("Answer format: {word: [most similar word, cosine similarity]}")
print(wordDict)

Answer format: {word: [most similar word, cosine similarity]}
{'dog': ['dogs', 0.81368625], 'whale': ['whales', 0.7918058], 'before': ['after', 0.8931248], 'however': ['although', 0.9336163], 'fabricate': ['fabricating', 0.61840194]}


## 2.2 Finding Analogies
Use vector addition and subtraction to compute target vectors for the analogies below. After computing each target vector, find the top three candidates by cosine similarity. Report the candidates and their similarities to the target vector.

- dog : puppy :: cat : ?:
- speak : speaker :: sing : ?:
- France : French :: England : ?:
- France : wine :: England : ?

In [ ]:
# dog-puppy+cat
wordList = [["dog", "puppy", "cat"], ["speak", "speaker", "sing"], ["france", "french", "england"], ["france", "wine", "england"]]

vecKeys = wv_from_bin.index_to_key
for analogy in wordList:
  targetVec = wv_from_bin[analogy[0]] - wv_from_bin[analogy[1]] + wv_from_bin[analogy[2]]
  
  candidateList = []
  similarityList = []
  for i in range(len(wv_from_bin)):
    cosine = np.dot(targetVec,wv_from_bin[i])/(norm(targetVec)*norm(wv_from_bin[i]))
    curword = vecKeys[i]
    
    if len(similarityList) < 3 or (len(similarityList) >= 3 and cosine >= similarityList[2]):
      similarityList.append(cosine) 
      similarityList.sort(reverse=True)
      candidateList.insert(similarityList.index(cosine), curword)
  print(f"{analogy[0]} : {analogy[1]} :: {analogy[2]} : ?:", candidateList[0:3])
  print("corresponding similarities:", similarityList[0:3])
# targetVec = wv_from_bin["dog"] - wv_from_bin["puppy"] + wv_from_bin["cat"]
# print("kitty", np.dot(targetVec,wv_from_bin["kitty"])/(norm(targetVec)*norm(wv_from_bin["kitty"])))
# print("kitten", np.dot(targetVec,wv_from_bin["kitten"])/(norm(targetVec)*norm(wv_from_bin["kitten"])))
# print("dogs", np.dot(targetVec,wv_from_bin["dogs"])/(norm(targetVec)*norm(wv_from_bin["dogs"])))

dog : puppy :: cat : ?: ['cat', 'dog', 'dogs']
corresponding similarities: [0.8246017, 0.745641, 0.6278181]
speak : speaker :: sing : ?: ['sing', 'speak', 'singing']
corresponding similarities: [0.7365669, 0.55140996, 0.548521]
france : french :: england : ?: ['england', 'wales', 'ireland']
corresponding similarities: [0.837254, 0.6498687, 0.6262423]
france : wine :: england : ?: ['england', 'france', 'britain']
corresponding similarities: [0.69291466, 0.6897908, 0.61666137]


# 3 Sentiment analysis

In [ ]:
!mkdir -p data/classification
!wget -O data/classification/train.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/train.txt
!wget -O data/classification/dev.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/dev.txt
!wget -O data/classification/test-blind.txt https://raw.githubusercontent.com/ranpox/comp3361-spring2024/main/assignments/A1/data/classification/test-blind.txt

The syntax of the command is incorrect.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.
'wget' is not recognized as an internal or external command,
operable program or batch file.


## 3.1 Logistic Regression

### Unigram Features

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, f1_score

# Read the training data
train_data = []
train_labels = []
with open('data/classification/train.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        sentiment = line[0]
        train_data.append(' '.join(line[1:]))
        train_labels.append(line[0])

# Convert the text data into a matrix of token counts
vectorizer = CountVectorizer()
x = vectorizer.fit_transform(train_data)

# Fit the LogisticRegression model
logReg = LogisticRegression()
logReg.fit(x, train_labels)

# Predict the labels for the test data
test_data = []
test_labels = []
with open('data/classification/dev.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        test_labels.append(line[0])
        test_data.append(' '.join(line[1:]))
test_data_transformed = vectorizer.transform(test_data)
predictions = logReg.predict(test_data_transformed)

print("unigram precision: ", precision_score(test_labels, predictions, average='macro'))
print("unigram recall: ", recall_score(test_labels, predictions, average='macro'))
print("unigram f1 score: ", f1_score(test_labels, predictions, average='macro'))


unigram precision:  0.7893626002532714
unigram recall:  0.7885829754988634
unigram f1 score:  0.7887062789717657


### Bigram Features

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, f1_score

# Read the training data
train_data = []
train_labels = []
with open('data/classification/train.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        train_data.append(' '.join(line[1:]))
        train_labels.append(line[0])

# Convert the text data into a matrix of token counts
vectorizer = CountVectorizer(ngram_range=(2, 2))
x = vectorizer.fit_transform(train_data)

# Fit the LogisticRegression model
logReg = LogisticRegression()
logReg.fit(x, train_labels)

# Predict the labels for the test data
test_data = []
test_labels = []
with open('data/classification/dev.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        test_labels.append(line[0])
        test_data.append(' '.join(line[1:]))
test_data_transformed = vectorizer.transform(test_data)
predictions = logReg.predict(test_data_transformed)

print("bigram precision: ", precision_score(test_labels, predictions, average='macro'))
print("bigram recall: ", recall_score(test_labels, predictions, average='macro'))
print("bigram f1 score: ", f1_score(test_labels, predictions, average='macro'))

# print(x.shape)
# print(x)

bigram precision:  0.7213118279569892
bigram recall:  0.7166161488591395
bigram f1 score:  0.7159533898305085


### GloVe Features

In [1]:
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, f1_score

# Read the training data
train_data = []
train_labels = []
with open('data/classification/train.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        sentiment = line[0]
        words = line[1:]
        # print(len(line))
        # avgVec = []
        # avgVec = np.mean([wv_from_bin[word] for word in words if word in wv_from_bin.key_to_index], axis=0)
        temp = []
        for word in words:
            if word in wv_from_bin.key_to_index:
                # avgVec.append(wv_from_bin[word])
                # train_data.append(wv_from_bin[word])
                temp.append(wv_from_bin[word])
            #     train_data.append(wv_from_bin["<UNK>"])
        train_data.append(temp)
        # train_data = np.append(train_data, avgVec)
        # train_data.append(np.mean(avgVec, axis=0))
        # train_data.append(' '.join(line[1:]))
        train_labels.append(line[0])

# print(train_data)

# print(len(train_data[0]))
# print(np.array(train_data).shape)

# Convert the text data into a matrix of token counts
# vectorizer = CountVectorizer()
# x = vectorizer.fit_transform(train_data)

# Fit the LogisticRegression model
logReg = LogisticRegression()
logReg.fit(train_data, train_labels)

# Predict the labels for the test data
test_data = []
test_labels = []
with open('data/classification/dev.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        test_labels.append(line[0])
        test_data.append(' '.join(line[1:]))
# Convert the text data into a matrix of token counts
vectorizer = CountVectorizer()
test_data_transformed = vectorizer.transform(test_data)
predictions = logReg.predict(test_data_transformed)

print("bigram precision: ", precision_score(test_labels, predictions, average='macro'))
print("bigram recall: ", recall_score(test_labels, predictions, average='macro'))
print("bigram f1 score: ", f1_score(test_labels, predictions, average='macro'))


NameError: name 'wv_from_bin' is not defined

Compare the performance of three types of features on dev set. Report the weighted average precision, recall and F1-score for each feature set.

| Feature | precision | recall | F1-score |
| ----------- | --------- | ------ | -------- |
| unigram     |0.7893626002532714|0.7885829754988634|0.7887062789717657|
| bigram      |0.7213118279569892|0.7166161488591395|0.7159533898305085|
| GloVe       |           |        |          |

## 3.2 Better Feature

In [80]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, recall_score, f1_score
from gensim.parsing.preprocessing import remove_stopwords, stem_text, strip_punctuation, strip_multiple_whitespaces


# Read the training data
train_data = []
train_labels = []
with open('data/classification/train.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        sentiment = line[0]
        words = ' '.join(line[1:])
        train_data.append((strip_multiple_whitespaces(strip_punctuation(words.replace('-RRB-', '').replace('-LRB-', '').lower()))))
        train_labels.append(line[0])

# print(train_data)
# Convert the text data into a matrix of token counts
vectorizer = CountVectorizer(ngram_range=(1, 4))
x = vectorizer.fit_transform(train_data)

# Fit the LogisticRegression model
logReg = LogisticRegression()
logReg.fit(x, train_labels)

# # Predict the labels for the test data
# test_data = []
# test_labels = []
# with open('data/classification/dev.txt', 'r', encoding="utf-8") as f:
#     for line in f:
#         line = line.split()
#         test_labels.append(line[0])
#         test_data.append(' '.join(line[1:]))
# test_data_transformed = vectorizer.transform(test_data)
# predictions = logReg.predict(test_data_transformed)

# print("precision: ", precision_score(test_labels, predictions, average='macro'))
# print("recall: ", recall_score(test_labels, predictions, average='macro'))
# print("f1 score: ", f1_score(test_labels, predictions, average='macro'))

test_data = []
with open('data/classification/test-blind.txt', 'r', encoding="utf-8") as f:
    for line in f:
        line = line.split()
        test_data.append(' '.join(line[1:]))
test_data_transformed = vectorizer.transform(test_data)
predictions = logReg.predict(test_data_transformed).tolist()

with open('data/classification/3035927116.test.txt', 'w', encoding="utf-8") as f:
    with open('data/classification/test-blind.txt', 'r', encoding="utf-8") as f2:
        for line in f2:
            f.write(predictions.pop(0) + "	" + line)